source/opt/fold.cpp

```diff
@@ -289,5 +290,54 @@ bool IsFoldableConstant(const analysis::Constant* cst) {
     return cst->AsNullConstant() != nullptr;
 }
 
+ir::Instruction* FoldInstruction(
+    ir::Instruction* inst,
+    std::function<uint32_t(uint32_t)> id_map ) {
+  if (!IsFoldableOpcode(inst->opcode())) {
+    return nullptr;
+  }
+
+  ir::IRContext* context = inst->context();
+  analysis::ConstantManager* const_mgr = context->get_constant_mgr();
+
+  std::vector<const analysis::Constant*> constants;
+  bool missing_constants = false;
+  inst->ForEachInId(
+      [&constants, &missing_constants, const_mgr, &id_map](uint32_t* op_id) {
+        uint32_t id = id_map(*op_id);
+        const analysis::Constant* const_op =
+            const_mgr->FindDeclaredConstant(id);
+        if (!const_op || !IsFoldableConstant(const_op)) {
+          constants.push_back(nullptr);
+          missing_constants = true;
+          return;
+        }
+        constants.push_back(const_op);
+      });
+
+  if (!missing_constants) {
+    uint32_t result_val = FoldScalars(inst->opcode(), constants);
+    const analysis::Constant* result_const =
+        const_mgr->GetConstant(const_mgr->GetType(inst), {result_val});
+    return const_mgr->GetDefiningInstruction(result_const);
+  }
+
+  // Check for constant parameters.
```

Comment:
This comment is a bit confusing.

I'd add a TODO, e.g. 
`// TODO: Other folding cases, such as arithmetic and logic absorption.`

Reply:
Done.